modules/tinymce/src/core/main/ts/fmt/FormatChanged.ts

```diff
@@ -5,126 +5,203 @@
+  readonly others: CustomCallback[];
 }
 
 const setup = (registeredFormatListeners: Cell<RegisteredFormats>, editor: Editor) => {
-  const currentFormats = Cell<Record<string, FormatChangeCallback[]>>({ });
-
   registeredFormatListeners.set({});
 
   editor.on('NodeChange', (e) => {
-    updateAndFireChangeCallbacks(editor, e.element, currentFormats, registeredFormatListeners.get());
+    updateAndFireChangeCallbacks(editor, e.element, registeredFormatListeners.get());
   });
 };
 
-const updateAndFireChangeCallbacks = (editor: Editor, elm: Element, currentFormats: Cell<FormatCallbacks>, formatChangeData: RegisteredFormats) => {
-  const formatsList = Obj.keys(currentFormats.get());
-  const newFormats: FormatCallbacks = { };
-  const matchedFormats: FormatCallbacks = { };
-
-  // Ignore bogus nodes like the <a> tag created by moveStart()
-  const parents = Arr.filter(FormatUtils.getParents(editor.dom, elm), (node) => node.nodeType === 1 && !node.getAttribute('data-mce-bogus'));
-
-  // Check for new formats
-  Obj.each(formatChangeData, (data: FormatData, format: string) => {
-    Tools.each(parents, (node: Node) => {
-      if (editor.formatter.matchNode(node, format, {}, data.similar)) {
-        if (formatsList.indexOf(format) === -1) {
-          // Execute callbacks
-          Arr.each(data.callbacks, (callback: FormatChangeCallback) => {
-            callback(true, { node, format, parents });
-          });
-
-          newFormats[format] = data.callbacks;
-        }
-
-        matchedFormats[format] = data.callbacks;
-        return false;
-      }
-
+const matchingNode = (editor: Editor, parents: Element[], format: string, similar: boolean, vars?: FormatVars): Optional<Element> =>
+  Arr.findUntil(
+    parents,
+    (node) => { // is this the node we want?
+      const matchingFormat = editor.formatter.matchNode(node, format, vars ?? {}, similar);
+      return !Type.isUndefined(matchingFormat);
+    },
+    (node) => { // Can we guarantee that no parent of this will be the node we want?
       if (MatchFormat.matchesUnInheritedFormatSelector(editor, node, format)) {
-        return false;
+        return true;
+      } else {
+        if (!similar) {
+          // If we want to find an exact match, then finding a similar match halfway up the parents tree is bad
+          return Type.isNonNullable(editor.formatter.matchNode(node, format, vars, true));
+        } else {
+          return false;
+        }
       }
-    });
-  });
-
-  // Check if current formats still match
-  const remainingFormats = filterRemainingFormats(currentFormats.get(), matchedFormats, elm, parents);
+    }
+  );
 
-  // Update the current formats
-  currentFormats.set({
-    ...newFormats,
-    ...remainingFormats
-  });
+const getParents = (editor: Editor, elm?: Element): Element[] => {
+  const element = elm ?? editor.selection.getNode();
+  return Arr.filter(FormatUtils.getParents(editor.dom, element), (node) =>
```

Comment:
Could `NodeType.isBogus` be used here? It will match on `data-mce-bogus=""` though but I think that is correct since it's an internal attribute. Anything marked with a bogus attribute should be bogus.